libshortfin/src/shortfin/local/scope.cc

```diff
@@ -82,6 +88,40 @@ std::vector<std::string_view> Scope::device_names() const {
   return names;
 }
 
+Program Scope::LoadUnboundProgram(std::span<const ProgramModule> modules,
+                                  Program::Options options) {
+  std::vector<iree_vm_module_t *> all_modules;
+  std::vector<iree_hal_device_t *> raw_devices;
+
+  // By default, bind all devices in the scope in order to the program.
+  for (Device *d : devices_) {
+    raw_devices.push_back(d->hal_device());
+  }
+
+  // Add a HAL module.
+  iree::vm_module_ptr hal_module;
+  SHORTFIN_THROW_IF_ERROR(iree_hal_module_create(
```

Comment:
fyi: at some point you may want to change this to something similar to what the tooling does in `iree_tooling_resolve_modules` - it uses `iree_vm_module_enumerate_dependencies` to walk the dependencies and add the required modules only as needed. to start you could use it just to see if the hal is used, but as you add other module types for exposing sharkfin functionality (or module versions; `iree_vm_module_dependency_t` has the minimum version required so you can switch between them, and whether they are optional/required) you'll probably want to build around that

Reply:
Thanks. Converted to a TODO.